S26-documentation/12-non-breaking-space.t

```diff
@@ -1,163 +1,86 @@
-    for @nbchars -> $nbc {
-        # first column
-        $code ~= "Raku";
-        $code ~= $nbc.chr;
-        $code ~= "Language";
-        for @bchars -> $c {
-            $code ~= $c.chr;
-        }
-        $code ~= "is the best!";
-
-        # second column
-        $code ~= ' | ';
-        $code ~= "Raku ";
-        $code ~= $nbc.chr;
-        $code ~= " Language";
-        for @bchars -> $c {
-            $code ~= $c.chr;
-        }
-        $code ~= "is the best!\n";
-    }
-    $code ~= "=end table\n";
-
-    my $n = +@nbchars;
-    # the planned num of tests is 1 + $n
-    $code ~= "plan 1 + $n * 2;\n";
-    $code ~= "my \$n = $n;\n";
-    $code ~= "my \$SPACE = ' ';\n";
-
-    $code ~= q:to/HERE/;
-        my $r = $=pod[0];
-        is $r.contents.elems, $n, "table has $n elements (rows)";
-        HERE
-
-    # create separate sub-tests for each non-breaking space char
-    for 0..^$n -> $i {
-        $code ~= "\n\{\n";
-        $code ~= "my \$i = $i;\n";
-        $code ~= "my \$row = $i + 1;\n";
-        $code ~= "my \$nbspc = \"{@nbchars[$i].chr}\";\n";
-
-        $code ~= q:to/HERE/;
-            my $res0 = "Raku{$nbspc}Language is the best!";
-            my $res1 = "Raku {$nbspc} Language is the best!";
-            # show cell chars separated by pipes
-            =begin comment
-            my $c0 = $r.contents[$i][0].comb.join('|');
-            my $r0 = $res0.comb.join('|');
-            my $c1 = $r.contents[$i][1].comb.join('|');
-            my $r1 = $res1.comb.join('|');
-            =end comment
-            my $c0 = show-space-chars($r.contents[$i][0]);
-            my $r0 = show-space-chars($res0);
-            my $c1 = show-space-chars($r.contents[$i][1]);
-            my $r1 = show-space-chars($res1);
-
-            is $r.contents[$i][0], $res0, "row $row, col 1: '$c0' vs '$r0'";
-            is $r.contents[$i][1], $res1, "row $row, col 2: '$c1' vs '$r1'";
-            HERE
-
-        $code ~= "}\n";
-    }
-
-    $code ~= q:to/HERE/;
-        ##### some subroutines for the EVALed file #####
-        sub int2hexstr($int, :$plain) {
-            # given an int, convert it to a hex string
-            if !$plain {
-                return sprintf("0x%04X", $int);
-            }
-            else {
-                return sprintf("%X", $int);
-            }
-        }
+use lib $?FILE.IO.parent(2).add('packages/Test-Helpers/lib');
+use Test::Misc :int2hexstr, :show-space-chars;
+
+plan 9;
+
+# two-column table with various whitespace chars
+=begin table
+Raku Language	  ᠎            　is the best! | Raku   Language	  ᠎            　is the best!
+Raku Language	  ᠎            　is the best! | Raku   Language	  ᠎            　is the best!
+Raku⁠Language	  ᠎            　is the best! | Raku ⁠ Language	  ᠎            　is the best!
+Raku﻿Language	  ᠎            　is the best! | Raku ﻿ Language	  ᠎            　is the best!
+=end table
+
+my $n = 4;
+my $SPACE = ' ';
+my $r = $=pod[0];
```

Comment:
With the previous comment in mind, this line would then be replaced with `my $r = get-pod;` and  the rest of the test would remain unchanged. So, we could get the best of two variants: list of chars with codes and names from the initial `EVAL`ed version, and easy to work with test code outside of any `EVAL`.

Reply:
I'll take another look, but, at the moment, I don't see how the EVAL block will be much smaller (except for the refactored subs).

Reply:
The point is not in making it smaller but to have only the pod section evaled. The biggest problem with the initial approach is the difficulty to analyze and modify code spread across several `~=` statements.

Simply put, from the original test I'm interested in preserving lines 15 to 79.

Reply:
I can use your example but I can't see how the generate the code internally in the same comp unit without more EVALing. I can generate it fine from another program.

Suggesstions? If you know how to do it please show me a short example.

Reply:
If I got the misunderstanding correctly, I don't mean to get rid of `EVAL` altogether. I only want to minimize its use. Thus, `EVAL`ing the generated code, as in my example I eval a piece of pod with a primitive sub, is totally ok.

I mean the separation I'm trying to reach is to make the test code, which is to be read and understood by a human, available directly in one clean piece; and only generate and `EVAL` the pod block as it is just test input data and barely would be modified or analyzed by a person.

Reply:
Yes, I understand what you want, but I don't see how to do it with a single script. Let me try a new tack and see your reaction. I should get it here later today.

Reply:
Let me repeat the example code here:

```raku
sub get-pod {
    my $pod;
    my sub pull-pod($p) {
        $pod = $p;
    }
    EVAL q:to/TEST/;
        =begin pod
        Pod text goes here...
        =end pod
        pull-pod($=pod[0]);
        TEST
        
    $pod
}
my $pod = get-pod;
isa-ok $pod, Pod::Block, "fine";
```

This approximately what I'd like to achieve.

Reply:
I looked again at the situation this morning but I won't start work until you're happy with my proposal below.


First note that I found at least eight test generation programs in **roast** and that is what I intend to add to the **S26** directory. Essentially, I will modify what I did to generate the current test by adding my Raku program and incorporate your excellent method to minimize the EVALed code.

### PROPOSAL

+ Create two code templates: 

    - test-file.raku.top
    - test-file.raku.bottom

+ Create the Raku test generator to:

    - write the  top part to the test file
    - generate and append the EVALed test code to the test file
    - append the bottom part to the test file (I will add some code to all parts as a check on planned tests such how many table rows are expected and how many were actually found)

+ Create a **README** for the directory to explain the situation

+ Put all three non-test parts in a new **tools** subdirectory

Reply:
I don't think two extra files and manipulations with them would greatly improve the test.

I probably failed to explain my idea clearly after all. Please, review 4fffc73 and check if it does what it should. I have grouped the tests into subtests. And added two more tests per row (per subtest). So, now, instead of just outputting the converted lines of ws codes it compares them.